xfeat/cat_encoder/_target_encoder.py

```diff
@@ -398,6 +398,7 @@ def fit(self, X: CSeries, y: CSeries):
         """
         # Label encoding if necessary
         if not cupy.can_cast(X.dtype, cupy.int):
+            X = X.to_array()
```

Comment:
Would you provide test cases that this line is necessary? I guess handling categorical columns for target encoding is covered in this test case. https://github.com/pfnet-research/xfeat/blob/master/tests/xfeat/cat_encoder/test_target_encoder.py#L21-L62

Reply:
I uploaded the test script for this PR: https://gist.github.com/smly/52e634e98a4cd5bca07cf72d979984c5
Unfortunately, I can't find any problem in handling categorical columns. I tested with `cudf 0.13.0` and `pandas 0.25.3`. Your problem might be resolved by https://github.com/pfnet-research/xfeat/pull/4.

If your problem is still not resolved, I hope you provide more detail such as version info, code example, and error messages.

Reply:
I'm using cudf `'0.15.0a+3896.g36bb50ac9'` the latest build as of today.

This is the code to reproduce the issue with passing object columns:
```
import cudf
import xfeat
df = cudf.DataFrame({'a': ["1", "2", "3"],
                    'b': [14, 20, 38],
                    'c': [100, 300, 400]})

df['a'] = df['a'].astype("object")
df.dtypes
# a    object
# b     int64
# c     int64
# dtype: object

# Trying with Pandas dataframe. This WORKS!
df_pandas = df.to_pandas()
df_ = xfeat.Pipeline([
    xfeat.SelectCategorical(), xfeat.LabelEncoder(output_suffix="")]
).fit_transform(
    df_pandas
)


# Now, using categorical encoder on cudf DF
df_ = xfeat.Pipeline([
    xfeat.SelectCategorical(), xfeat.LabelEncoder(output_suffix="")]
).fit_transform(
    df
)

# Results in the Error below
```


```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-076ab16be512> in <module>
      2     xfeat.SelectCategorical(), xfeat.LabelEncoder(output_suffix="")]
      3 ).fit_transform(
----> 4     df
      5 )

/opt/conda/envs/rapids/lib/python3.6/site-packages/xfeat-0.1.0-py3.6.egg/xfeat/pipeline/_pipeline.py in fit_transform(self, input_df)
     57         """
     58         for transform in self._transforms:
---> 59             input_df = transform.fit_transform(input_df)
     60         return input_df
     61 

/opt/conda/envs/rapids/lib/python3.6/site-packages/xfeat-0.1.0-py3.6.egg/xfeat/cat_encoder/_basic_encoder.py in fit_transform(self, input_df)
    103         """
    104         self.fit(input_df)
--> 105         return self.transform(input_df)
    106 
    107     def transform(self, input_df: XDataFrame) -> XDataFrame:

/opt/conda/envs/rapids/lib/python3.6/site-packages/xfeat-0.1.0-py3.6.egg/xfeat/cat_encoder/_basic_encoder.py in transform(self, input_df)
    125         for col in input_cols:
    126             out_col = self._output_prefix + col + self._output_suffix
--> 127             X = self._uniques[col].get_indexer(new_df[col])
    128 
    129             if self._unseen == "n_unique":

/opt/conda/envs/rapids/lib/python3.6/site-packages/pandas/core/indexes/base.py in get_indexer(self, target, method, limit, tolerance)
   2707     def get_indexer(self, target, method=None, limit=None, tolerance=None):
   2708         method = missing.clean_reindex_fill_method(method)
-> 2709         target = ensure_index(target)
   2710         if tolerance is not None:
   2711             tolerance = self._convert_tolerance(tolerance, target)

/opt/conda/envs/rapids/lib/python3.6/site-packages/pandas/core/indexes/base.py in ensure_index(index_like, copy)
   5329         return index_like
   5330     if hasattr(index_like, "name"):
-> 5331         return Index(index_like, name=index_like.name, copy=copy)
   5332 
   5333     if is_iterator(index_like):

/opt/conda/envs/rapids/lib/python3.6/site-packages/pandas/core/indexes/base.py in __new__(cls, data, dtype, copy, name, tupleize_cols, **kwargs)
    418 
    419         elif hasattr(data, "__array__"):
--> 420             return Index(np.asarray(data), dtype=dtype, copy=copy, name=name, **kwargs)
    421         elif data is None or is_scalar(data):
    422             raise cls._scalar_data_error(data)

/opt/conda/envs/rapids/lib/python3.6/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
     83 
     84     """
---> 85     return array(a, dtype, copy=False, order=order)
     86 
     87 

/opt/conda/envs/rapids/lib/python3.6/site-packages/cudf/core/frame.py in __array__(self, dtype)
   1635     def __array__(self, dtype=None):
   1636         raise TypeError(
-> 1637             "Implicit conversion to a host NumPy array via __array__ is not "
   1638             "allowed, To explicitly construct a GPU array, consider using "
   1639             "cupy.asarray(...)\nTo explicitly construct a "

TypeError: Implicit conversion to a host NumPy array via __array__ is not allowed, To explicitly construct a GPU array, consider using cupy.asarray(...)
To explicitly construct a host array, consider using .to_array()
```

Reply:
I'll work the the PR more, I see what error you were talking about.

Reply:
Thanks for the detailed info. I had not considered supporting the nightly version of `cudf`, but it's valuable to support it for improving compatibility. I will check tests with `cudf 0.14` and `cudf 0.15`.

Reply:
Thanks! Let me know if the PR fixes the issues. All the current xfeat tests passed, let me know if I've missed something!